proxy/src/bind.rs

```diff
@@ -323,36 +339,75 @@ where
 // ===== impl Binding =====
 
-impl<B: tower_h2::Body + 'static> tower::Service for Binding<B> {
+impl<B: tower_h2::Body + 'static> tower::Service for BoundService<B> {
     type Request = <Stack<B> as tower::Service>::Request;
     type Response = <Stack<B> as tower::Service>::Response;
     type Error = <Stack<B> as tower::Service>::Error;
     type Future = <Stack<B> as tower::Service>::Future;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        match *self {
+        let ready = match self.binding {
             // A service is already bound, so poll its readiness.
             Binding::Bound(ref mut svc) |
-            Binding::BindsPerRequest { next: Some(ref mut svc), .. } => svc.poll_ready(),
+            Binding::BindsPerRequest { next: Some(ref mut svc) } => svc.poll_ready(),
 
             // If no stack has been bound, bind it now so that its readiness can be
             // checked. Store it so it can be consumed to dispatch the next request.
-            Binding::BindsPerRequest { ref endpoint, ref protocol, ref bind, ref mut next } => {
-                let mut svc = bind.bind_stack(endpoint, protocol);
-                let ready = svc.poll_ready()?;
+            Binding::BindsPerRequest { ref mut next } => {
+                let mut svc = self.bind.bind_stack(&self.endpoint, &self.protocol);
+                let ready = svc.poll_ready();
                 *next = Some(svc);
-                Ok(ready)
+                ready
             }
+        };
+
+        // If there was a connect error, don't terminate this BoundService
+        // completely. Instead, simply clean up the inner service, prepare to
+        // make a new one, and tell our caller that we could maybe be ready
+        // if they call `poll_ready` again.
+        //
+        // If they *don't* call `poll_ready` again, that's ok, we won't ever
+        // try to connect again.
+        match ready {
+            Err(ReconnectError::Connect(err)) => {
+                error!("connect error to {:?}: {}", self.endpoint, err);
+                match self.binding {
+                    Binding::Bound(ref mut svc) => {
+                        *svc = self.bind.bind_stack(&self.endpoint, &self.protocol);
+                    },
+                    Binding::BindsPerRequest { ref mut next } => {
+                        next.take();
+                    }
+                }
+
+                // So, this service isn't "ready" now, but to only make the
+                // service try another connect if there are still requests
+                // waiting, this doesn't poll the newly bound service.
+                //
+                // But, to return NotReady, we must notify the task. So,
+                // this notifies the task immediately, and figures that
+                // whoever owns this service will call `poll_ready` if they
+                // are still interested.
+                task::current().notify();
+                Ok(Async::NotReady)
```

Comment:
What are the merits of doing this versus looping with the new state?

Reply:
I tried to make that clear in the comment, so it seems I should add to it.

If we loop in here, then we will be eagerly setting up a new connection, even if the buffer wrapping this service determines that it's queue of requests have run out. This instead allows the buffer to determine if it should loop or not.

Reply:
thanks.

Reply:
I think that first sentence in the comment just needs to be rearranged a bit
```
// This service isn't ready yet. Instead of trying to make it ready,
// schedule the task for notification so that the caller can
// determine whether readiness is still necessary (i.e. whether
// there are still requests to be sent).
```

... or something like that...